hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -808,10 +807,17 @@ private ClientConnection getConnectionFromLoadBalancer(boolean dataMember) {
             return null;
         }
 
-        return getConnection(member.getUuid());
+        return activeConnections.get(member.getUuid());
     }
 
-    private void authenticateOnCluster(TcpClientConnection connection) {
+    /**
```

Comment:
you gave extra responsibility to the method `authenticateOnCluster`, rather than simply just authentication. Hence, you needed to put this explanation for the implicit returned connection. I would prefer this method to just do authentication and do existing connection check at `getOrConnectToAddress`. you also can close the connection at the same method rather than at `handleSuccessfulAuth `. This is more understandable to me, separation of concerns.

Reply:
We are getting a lock at `handleSuccessfulAuth`. The check if the connection already exists should happen under that lock together with adding to `activeConnections` map if the connection did not exist. Not sure if I can refactor any further.

Reply:
you can still do the check there at line https://github.com/hazelcast/hazelcast/pull/17844/files#diff-27e248511ddd6e03d414477f847d9c1b05ab33c329e40d318345cb2ed0d0ead2L874  where you can do `putIfAbsent` and if put fails, it can throw exception meaning that the connection auth failed due to an already existing connection and you do not even need a second check again.